### PR TITLE
 make cadvisor metrics set configurable in kubelet

### PR DESCRIPTION
**Pull request by KielChan**
Signed-off-by: ChenQingya <qingya.chen520@gmail.com>

**What type of PR is this?**
> /kind feature

**What this PR does / why we need it**:

Add `CustomCadvisorMetrics` feature gate for kubelet.

**Which issue(s) this PR fixes**:

Fixes # none

**Special notes for your reviewer**:  I have issued a PR before, and comments here: https://github.com/kubernetes/kubernetes/pull/88841

**Does this PR introduce a user-facing change?**:  Yes, it will change metrics exposed by kubelet API.

```release-note

```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**: https://github.com/kubernetes/enhancements/issues/1920

```docs

```

### CODE REVIEW

File: pkg/kubelet/cadvisor/util_test.go [hunk 1]
```diff
@@ -55,3 +56,44 @@ func TestCapacityFromMachineInfoWithHugePagesEnable(t *testing.T) {
 func TestCrioSocket(t *testing.T) {
 	assert.EqualValues(t, CrioSocket, crio.CrioSocket, "CrioSocket in this package must equal the one in github.com/google/cadvisor/container/crio/client.go")
 }
+
+func TestDefaultCadvisorMetricSet(t *testing.T) {
+	usingLegacyStatsTable := []bool{false, true}
```
Comment (kikisdeliveryservice) [inline pkg/kubelet/cadvisor/util_test.go:61]: anything for `AcceleratorUsageMetrics`?

File: pkg/kubelet/cadvisor/util_test.go [hunk 2]
```diff
@@ -55,3 +56,44 @@ func TestCapacityFromMachineInfoWithHugePagesEnable(t *testing.T) {
...
+	for idx, usingLegacyStats := range usingLegacyStatsTable {
+		actual := DefaultCadvisorMetricSet(usingLegacyStats)
+		if reflect.DeepEqual(actual, expected[idx]) {
+			t.Errorf("test default metric set for cadvisor error, expected: %+v, actual: %+v", expected[idx], actual)
+		}
+	}
+}
+
+func TestCustomCadvisorMetricSet(t *testing.T) {
+	usingLegacyStatsTable := []bool{false, true}
```
Comment (kikisdeliveryservice) [inline pkg/kubelet/cadvisor/util_test.go:79]: does this test cover the check on `kubefeatures.DisableAcceleratorUsageMetrics`?

**Review by kikisdeliveryservice — COMMENTED** [review]
Two small questions

File: pkg/kubelet/cadvisor/util_test.go (hunk 1 above)
Comment (KielChan) [reply pkg/kubelet/cadvisor/util_test.go:61]: Fine, I will add it today.

Comment (kikisdeliveryservice) [reply pkg/kubelet/cadvisor/util_test.go:61]: thank you, the updates look good!

**Comment by KielChan** [comment]
maybe something wrong with e2e test, I will fix it today.

**Comment by dashpole** [comment]
It looks like this is based on a KEP: https://github.com/kubernetes/enhancements/pull/1921, so we will need to wait for that to be merged as implementable before this can proceed

**Comment by KielChan** [comment]
anything wrong with gcs? :(

**Comment by BenTheElder** [comment]
>  F0813 08:55:05.347489    2383 kubelet.go:1296] Failed to start ContainerManager failed to get rootfs info: cannot find filesystem info for device "/dev/sda1"

<panic>

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/93650/pull-kubernetes-e2e-gce-alpha-features/1293827136741707776/artifacts/e2e-a2bfd0c6c9-cc220-master/kubelet.log

**Comment by SergeyKanzhelev** [comment]
/assign @bobbypage

FYI

**Comment by ehashman** [comment]
@warmchang please don't remove stale labels without an accompanying update to the PR. This PR is not mergeable in its current state.

@KielChan feel free to reopen this once you've rebased if you get a chance to work on this!

/close

**Comment by yuzhiquan** [comment]
ping @KielChan for move on

**Comment by yuzhiquan** [comment]
@ehashman we need this feature in our cluster, so if @KielChan not respond, my team would like to working on this, WDYT？

**Comment by SergeyKanzhelev** [comment]
we need to design changes to cadvisor endpoint carefully as long term many metrics would be moving to cri stats. Also consider using standalone cAdvisor. It may be easier to implement configuration feature, for this I'd advice to start conversation at cadvisor repository

**Pull request closed without merging**